### ISSUE TITLE
SecretStoreException::UnknownException "Error while trying to load the Logstash keystore"

### ISSUE THREAD

**Issue by sebwas** [elastic/stack-docker#52]
Hello

Trying to bring up logstash after the installation went well brings up the following error:

```
ERROR: Failed to load settings file from "path.settings". Aborting... path.setting=/usr/share/logstash/config, exception=Java::OrgLogstashSecretStore::SecretStoreException::UnknownException, message=>Error while trying to load the Logstash keystore
[ERROR] 2018-08-02 14:47:36.244 [main] Logstash - java.lang.IllegalStateException: Logstash stopped processing because of an error: (SystemExit) exit
```

I have no idea where to start looking.

I'm running it on Windows and the rest of the containers boot up fine.

Is there any way I can try to access the keystore manually to see if it is corrupt or anything?

Thank you!

**Comment by fxdgear** [comment]
the keystore is created in `./config/logstash` you can inspect it there. 

it's possible it was corrupt in it's creation. Are you running windows 10/8/7? I've only tested this on 10. Though when trying to get multiple systems working together it can sometimes have failures :/ 

It might be worth cleaning up the repo and removing all the keystores and trying again. I'd take a look at the `Makefile` and look at the `clean` target to learn what files/contaienrs/volumes to remove. 

If you need more assistance please reply :)

**Comment by sebwas** [comment]
Thank you for your reply.

I am using Windows 10.

I tried to clean everything up, and set it all up again (`docker-compose down -v --remove-orphans && git clean -xfd && docker-compose -f setup.yml up`).
The same error happens.

Then, I tried the following:
```
PS> docker-compose run --rm logstash bin/logstash-keystore list
```
which resulted in the same original error ("Failed to load settings file from "path.settings" ...") and additionally
`[ERROR] 2018-08-03 12:29:35.832 [main] secretstorecli - Error while trying to load the Logstash keystore {:cause=>java.nio.file.NoSuchFileException: /usr/share/logstash/config/logstash.keystore, :backtrace=> (... cut for brevity)`.

However, running `ls -l /usr/share/logstash/config/logstash.keystore` in the container reveals that indeed it exists: `-rwxrwxrwx 1 root root 746 Aug  3 12:19 /usr/share/logstash/config/logstash.keystore`.

So I'm none the wiser.

**Comment by fxdgear** [comment]
@sebwas thanks for this. 

I'm rebooting my laptop into windows now. I'm going to do some experimenting. 

in the meantime can you tell me about your setup?

* docker version
* disks
* compose version
* any permissions you might have on your disk
* anything that might be considered "non standard"? 

Thanks

**Comment by sebwas** [comment]
Looking forward to hearing your results.
```
PS C:\Users\sebas> docker version
Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:05:28 2018
 OS/Arch:           windows/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:13:46 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```
```
PS C:\Users\sebas> docker-compose version
docker-compose version 1.22.0, build f46880fe
docker-py version: 3.4.1
CPython version: 3.6.6
OpenSSL version: OpenSSL 1.0.2o  27 Mar 2018
```

I'm not 100% sure how to answer the disks / disk permission questions. Could you be more specific?
There should not be anything considered "non standard" though in my understanding of things.

Thank you

**Comment by fxdgear** [comment]
Thanks for your response. 

When it comes to "non-standard" stuff, I was hoping something would just stick out in your mind. Personally I'm not a windows user by day. I have a laptop that's dual booted so i can verify things work on Windows, but my daily driver is linux :) 

BUT I can verify that I just rm'd the git repo, re-cloned it. deleted every image/container/volume/network of my windows machine then I re-did the whole setup: 

First I ran: `docker-compose -f setup.yml up`. I waited until the final message shows up saying:

```
setup_1  | Setup completed successfully. To start the stack please run:
setup_1  | 	 docker-compose up -d
setup_1  |
setup_1  | If you wish to remove the setup containers please run:
setup_1  | 	docker-compose -f docker-compose.yml -f docker-compose.setup.yml down --remove-orphans
setup_1  |
setup_1  | You will have to re-start the stack after removing setup containers.
setup_1  |
setup_1  | Your 'elastic' user password is: <some_generated_password>
stack-docker_setup_1 exited with code 0
```
After this I opened my browser to http://localhost:5601 and was able to log in with the username `elastic` and the generated password. 

I then attempted to run `docker-compose run --rm logstash bin/logstash-keystore list` and it ran without failure. 

I also ran this long command: 
```
docker run --rm -it -v /c/Users/nick/develop/elastic/stack-docker/config/logstash/logstash.keystore:/user/share/logstash/config/logstash.keystore docker.elastic.co/logstash/logstash:6.3.0 bin/logstash-keystore list
```

The purpose here being to isolate just the keystore, and avoid any of the `docker-compose`-ness of the situation. But this too ran successfully for me. 

My question now is, does the container `stack-docker_setup_1` exit with code 0 or a different code?

**Comment by sebwas** [comment]
Thank you for putting your time into that.

I think my host is pretty standard regarding docker, with maybe the only exception being that additionally I use docker on WSL, too. For details see [this tutorial](https://blogs.technet.microsoft.com/virtualization/2017/12/08/wsl-interoperability-with-docker/). But I don't suspect this to have anything to do with it. (At least I can't see how)

My setup ran fine, every setup container exited with code 0, so that running the whole setup resulted in the same output as yours.

I am also able to log in to Kibana and the other containers seem to run well.

Really clueless by now.

Just to make sure: which branch are you using? `master`?

**Comment by fxdgear** [comment]
Yes, I'm running `master`. 

I'm not using WSL. I tried to use windows as my main driver about a year ago. But I found running WSL and Docker for Windows would choke my system down sooo bad. I went back to linux cause it was too painful. (lenovo x1 yoga, 16gb ram still wasn't enough to run WSL/D4W). 

Can you try removing WSL from the mix? Just use PowerShell?

If you try just running the `logstash` container without the docker-compose do you still get errors re-`logstash.keystore`?

**Comment by sebwas** [comment]
I understand that. I didn't use Windows before my hardware upgrade.

I am not using WSL for setting it up, this time, because path references are a PITA using this. I was just using Powershell for this.

Well, guess what ... Trying to restart everything today did actually work. The only difference is that I used another name for the submodule folder I was using (namely, the default one, `stack-docker` instead of `elastic-stack` that I had been using before). However I can't find any hardcoded references to the name of the path, so that should not have been the problem.

In conclusion I did not find a solution to the problem I was having. Just retrying yielded success.

:-/
